File: Advanced/PageAndRDL/CustomPdfExport/VB.NET/PdfRendering/Pdf/PdfSharpSettings.vb
Imports GrapeCity.ActiveReports.Extensibility.Rendering
Imports System.Globalization
Imports System.ComponentModel
Imports System.Collections.Specialized
Imports System
Imports GrapeCity.ActiveReports.PageReportModel

''' <summary>
''' PDF settings.
''' </summary>
Public NotInheritable Class PdfSharpSettings
		Implements ISettings

		''' <summary>
		''' Initializes a new instance of the <see cref="PdfSharpSettings"/> class.
		''' </summary>
		Public Sub New()
			EmbedFonts = True
			EndPage = -1
			Target = TargetDeviceKind.Export
		End Sub

		''' <summary>
		''' Initializes a new instance of the <see cref="Settings"/> class.
		''' </summary>
		Public Sub New(settings As NameValueCollection)
			Me.New()
			CType(Me, ISettings).ApplySettings(settings)
		End Sub



	''' <summary>
	''' Embeds fonts or not.
	''' </summary>
	<DefaultValue(True)>
	<PdfDisplayName("EmbedFonts")>
	<PdfDescription("EmbedFontsDesc")>
	Public Property EmbedFonts() As Boolean

	''' <summary>
	''' Start page number.
	''' </summary>
	<DefaultValue(0)>
	<PdfDisplayName("StartPage")>
	<PdfDescription("StartPageDesc")>
	Public Property StartPage() As Integer

	''' <summary>
	''' End page number.
	''' </summary>
	<DefaultValue(-1)>
	<PdfDisplayName("EndPage")>
	<PdfDescription("EndPageDesc")>
	Public Property EndPage() As Integer

		' just parsing
		Friend Property Target() As TargetDeviceKind

#Region "ISettings implementation"

		''' <summary>
		''' Returns a <see cref="T:System.Collections.Specialized.NameValueCollection" /> containing the settings for the rendering extension.
		''' </summary>        
		Function GetSettings() As NameValueCollection Implements ISettings.GetSettings
			Dim settings As New NameValueCollection()
			If Not EmbedFonts Then
				settings("EmbedFonts") = Boolean.FalseString
			End If
			If StartPage <> 0 Then
				settings("StartPage") = StartPage.ToString(CultureInfo.InvariantCulture)
			End If
			If EndPage <> -1 Then
				settings("EndPage") = EndPage.ToString(CultureInfo.InvariantCulture)
			End If
			If Target <> TargetDeviceKind.Export Then
				settings("Target") = Target.ToString()
			End If
			Return settings
		End Function

		''' <summary>
		''' Apply settings for the rendering extension.
		''' </summary>
		''' <param name="settings">Settings for the rendering extension.</param>
		Sub ApplySettings(settings As NameValueCollection) Implements ISettings.ApplySettings
			settings = If(settings, New NameValueCollection())
			EmbedFonts = String.IsNullOrEmpty(settings.[Get]("EmbedFonts")) OrElse settings.[Get]("EmbedFonts").Equals(Boolean.TrueString, StringComparison.OrdinalIgnoreCase)
			Dim page As Integer
			StartPage = If(Integer.TryParse(If(settings.[Get]("StartPage"), String.Empty), page), page, 0)
			EndPage = If(Integer.TryParse(If(settings.[Get]("EndPage"), String.Empty), page), page, -1)
			' just parsing
			Target = If(String.IsNullOrEmpty(settings.[Get]("Target")), TargetDeviceKind.Export, CType([Enum].Parse(GetType(TargetDeviceKind), settings.[Get]("Target"), True), TargetDeviceKind))
		End Sub

#End Region
	End Class

File: Advanced/PageAndRDL/CustomPdfExport/C#/PdfRendering/Pdf/PdfSharpSettings.cs
using System;
using System.Collections.Specialized;
using System.ComponentModel;
using System.Globalization;
using System.Reflection;
using System.Resources;
using GrapeCity.ActiveReports.Extensibility.Rendering;
using GrapeCity.ActiveReports.PageReportModel;

namespace GrapeCity.ActiveReports.Samples.Export.Rendering.Pdf
{
	/// <summary>
	/// PDF settings.
	/// </summary>
	public sealed class PdfSharpSettings : ISettings
	{
		private static readonly string _embedFonts = Resources.EmbedFonts;
		private  readonly string _startPage = Resources.StartPage;
		private  readonly string _endPage = Resources.EndPage;

		/// <summary>
		/// Initializes a new instance of the <see cref="PdfSharpSettings"/> class.
		/// </summary>
		public PdfSharpSettings()
		{
			EmbedFonts = true;
			EndPage = -1;
			Target = TargetDeviceKind.Export;
		}

		/// <summary>
		/// Initializes a new instance of the <see cref="Settings"/> class.
		/// </summary>
		public PdfSharpSettings(NameValueCollection settings) : this()
		{
			((ISettings)this).ApplySettings(settings);
		}

		/// <summary>
		/// Embeds fonts or not.
		/// </summary>
		[DefaultValue(true)]
		[PdfDisplayName("EmbedFonts")]
		[PdfDescription("EmbedFontsDesc")]
		public bool EmbedFonts { get; set; }

		/// <summary>
		/// Start page number.
		/// </summary>
		[DefaultValue(0)]
		[PdfDisplayName("StartPage")]
		[PdfDescription("StartPageDesc")]
		public int StartPage { get; set; }

		/// <summary>
		/// End page number.
		/// </summary>
		[DefaultValue(-1)]
		[PdfDisplayName("EndPage")]
		[PdfDescription("EndPageDesc")]
		public int EndPage { get; set; }

		// just parsing
		internal TargetDeviceKind Target { get; private set; }

		#region ISettings implementation

		/// <summary>
		/// Returns a <see cref="T:System.Collections.Specialized.NameValueCollection" /> containing the settings for the rendering extension.
		/// </summary>        
		NameValueCollection ISettings.GetSettings()
		{
			var settings = new NameValueCollection();
			if (!EmbedFonts)
				settings["EmbedFonts"] = bool.FalseString;
			if (StartPage != 0)
				settings["StartPage"] = StartPage.ToString(CultureInfo.InvariantCulture);
			if (EndPage != -1)
				settings["EndPage"] = EndPage.ToString(CultureInfo.InvariantCulture);
			if (Target != TargetDeviceKind.Export)
				settings["Target"] = Target.ToString();
			return settings;
		}

		/// <summary>
		/// Apply settings for the rendering extension.
		/// </summary>
		/// <param name="settings">Settings for the rendering extension.</param>
		void ISettings.ApplySettings(NameValueCollection settings)
		{
			settings = settings ?? new NameValueCollection();
			EmbedFonts = string.IsNullOrEmpty(settings.Get("EmbedFonts")) || settings.Get("EmbedFonts").Equals(bool.TrueString, StringComparison.OrdinalIgnoreCase);
			int page;
			StartPage = int.TryParse(settings.Get("StartPage") ?? string.Empty, out page) ? page : 0;
			EndPage = int.TryParse(settings.Get("EndPage") ?? string.Empty, out page) ? page : -1;
			// just parsing
			Target = string.IsNullOrEmpty(settings.Get("Target")) ? TargetDeviceKind.Export : (TargetDeviceKind)Enum.Parse(typeof(TargetDeviceKind), settings.Get("Target"), true);
		}

		#endregion
	}
}
